File: contests/detective-pikaptcha/level2.cs
using System;
using System.Collections.Generic;

public class Pikachu
{
    public (int, int) start = (0, 0);
    public (int, int) previous = (0, 0);
    public int x = 0;
    public int y = 0;
    public string facing = "";
    public Dictionary<string, List<string>> priorities = new Dictionary<string, List<string>>();

    public bool IsTrapped()
    {
        return previous == (x, y);
    }
}

public class Game
{
    private const char WALL = '#';

    private const string FOLLOW_RIGHT = "R";
    private const string FOLLOW_LEFT = "L";

    private const string RIGHT = ">";
    private const string DOWN = "v";
    private const string LEFT = "<";
    private const string UP = "^";

    private readonly string[] DIRECTIONS = { RIGHT, DOWN, LEFT, UP };

    private List<string[]> rows = new List<string[]>();
    private int width = 0;
    private int height = 0;
    private Pikachu pika = new Pikachu();

    public void ReadInput()
    {
        string[] dimensions = Console.ReadLine().Split();
        width = int.Parse(dimensions[0]);
        height = int.Parse(dimensions[1]);

        for (int y = 0; y < height; y++)
        {
            string line = Console.ReadLine();
            rows.Add(new string[width]);
            for (int x = 0; x < width; x++)
            {
                if (Array.IndexOf(DIRECTIONS, line[x].ToString()) != -1)
                {
                    pika.start = (x, y);
                    pika.x = x;
                    pika.y = y;
                    pika.facing = line[x].ToString();
                }
                if (line[x] == WALL)
                {
                    rows[y][x] = WALL.ToString();
                }
                else
                {
                    rows[y][x] = "0";
                }
            }
        }

        string follow = Console.ReadLine();
        if (follow == FOLLOW_RIGHT)
        {
            pika.priorities = new Dictionary<string, List<string>>()
            {
                { UP, new List<string> { RIGHT, UP, LEFT, DOWN } },
                { DOWN, new List<string> { LEFT, DOWN, RIGHT, UP } },
                { RIGHT, new List<string> { DOWN, RIGHT, UP, LEFT } },
                { LEFT, new List<string> { UP, LEFT, DOWN, RIGHT } }
            };
        }
        else if (follow == FOLLOW_LEFT)
        {
            pika.priorities = new Dictionary<string, List<string>>()
            {
                { UP, new List<string> { LEFT, UP, RIGHT, DOWN } },
                { DOWN, new List<string> { RIGHT, DOWN, LEFT, UP } },
                { RIGHT, new List<string> { UP, RIGHT, DOWN, LEFT } },
                { LEFT, new List<string> { DOWN, LEFT, UP, RIGHT } }
            };
        }
    }

    public void Solve()
    {
        while (!IsPikachuBackAtTheStart())
        {
            List<string> priorityList = pika.priorities[pika.facing];
            pika.previous = (pika.x, pika.y);

            foreach (string priority in priorityList)
            {
                if (priority == RIGHT)
                {
                    if (PikachuCanGoRight())
                    {
                        pika.facing = RIGHT;
                        pika.x += 1;
                        break;
                    }
                }
                else if (priority == DOWN)
                {
                    if (PikachuCanGoDown())
                    {
                        pika.facing = DOWN;
                        pika.y += 1;
                        break;
                    }
                }
                else if (priority == LEFT)
                {
                    if (PikachuCanGoLeft())
                    {
                        pika.facing = LEFT;
                        pika.x -= 1;
                        break;
                    }
                }
                else if (priority == UP)
                {
                    if (PikachuCanGoUp())
                    {
                        pika.facing = UP;
                        pika.y -= 1;
                        break;
                    }
                }
            }

            if (pika.IsTrapped())
            {
                break;
            }
            rows[pika.y][pika.x] = (int.Parse(rows[pika.y][pika.x]) + 1).ToString();
        }
    }

    public bool IsPikachuBackAtTheStart()
    {
        return (pika.x, pika.y) == pika.start && rows[pika.y][pika.x] != "0";
    }

    public bool PikachuCanGoDown()
    {
        return pika.y < height - 1 && rows[pika.y + 1][pika.x] != WALL.ToString();
    }

    public bool PikachuCanGoLeft()
    {
        return pika.x > 0 && rows[pika.y][pika.x - 1] != WALL.ToString();
    }

    public bool PikachuCanGoRight()
    {
        return pika.x < width - 1 && rows[pika.y][pika.x + 1] != WALL.ToString();
    }

    public bool PikachuCanGoUp()
    {
        return pika.y > 0 && rows[pika.y - 1][pika.x] != WALL.ToString();
    }

    public void PrintSolution()
    {
        for (int y = 0; y < height; y++)
        {
            for (int x = 0; x < width; x++)
            {
                Console.Write(rows[y][x]);
            }
            Console.WriteLine();
        }
    }
}

class Program
{
    static void Main(string[] args)
    {
        Game game = new Game();
        game.ReadInput();
        game.Solve();
        game.PrintSolution();
    }
}

File: contests/detective-pikaptcha/level2.ts
class Pikachu {
  public start: [number, number] = [0, 0];
  public previous: [number, number] = [0, 0];
  public x: number = 0;
  public y: number = 0;
  public facing: string = "";
  public priorities: { [key: string]: string[] } = {};

  public isTrapped(): boolean {
    return this.previous[0] === this.x && this.previous[1] === this.y;
  }
}

class Game {
  private WALL: string = "#";

  private FOLLOW_RIGHT: string = "R";
  private FOLLOW_LEFT: string = "L";

  private RIGHT: string = ">";
  private DOWN: string = "v";
  private LEFT: string = "<";
  private UP: string = "^";

  private DIRECTIONS: string[] = [this.RIGHT, this.DOWN, this.LEFT, this.UP];

  private rows: string[][] = [];
  private width: number = 0;
  private height: number = 0;
  private pika: Pikachu = new Pikachu();

  public readInput(): void {
    const input = readline().split(" ");
    this.width = parseInt(input[0]);
    this.height = parseInt(input[1]);

    for (let y = 0; y < this.height; y++) {
      const line = readline();
      this.rows.push([]);
      for (let x = 0; x < this.width; x++) {
        if (this.DIRECTIONS.includes(line[x])) {
          this.pika.start = [x, y];
          this.pika.x = x;
          this.pika.y = y;
          this.pika.facing = line[x];
        }
        if (line[x] === this.WALL) {
          this.rows[y].push(this.WALL);
        } else {
          this.rows[y].push("0");
        }
      }
    }

    const follow = readline();
    if (follow === this.FOLLOW_RIGHT) {
      this.pika.priorities = {
        [this.UP]: [this.RIGHT, this.UP, this.LEFT, this.DOWN],
        [this.DOWN]: [this.LEFT, this.DOWN, this.RIGHT, this.UP],
        [this.RIGHT]: [this.DOWN, this.RIGHT, this.UP, this.LEFT],
        [this.LEFT]: [this.UP, this.LEFT, this.DOWN, this.RIGHT],
      };
    } else if (follow === this.FOLLOW_LEFT) {
      this.pika.priorities = {
        [this.UP]: [this.LEFT, this.UP, this.RIGHT, this.DOWN],
        [this.DOWN]: [this.RIGHT, this.DOWN, this.LEFT, this.UP],
        [this.RIGHT]: [this.UP, this.RIGHT, this.DOWN, this.LEFT],
        [this.LEFT]: [this.DOWN, this.LEFT, this.UP, this.RIGHT],
      };
    }
  }

  public solve(): void {
    while (!this.isPikachuBackAtTheStart()) {
      const priorityList = this.pika.priorities[this.pika.facing];
      this.pika.previous = [this.pika.x, this.pika.y];

      for (const priority of priorityList) {
        if (priority === this.RIGHT) {
          if (this.pikachuCanGoRight()) {
            this.pika.facing = this.RIGHT;
            this.pika.x += 1;
            break;
          }
        } else if (priority === this.DOWN) {
          if (this.pikachuCanGoDown()) {
            this.pika.facing = this.DOWN;
            this.pika.y += 1;
            break;
          }
        } else if (priority === this.LEFT) {
          if (this.pikachuCanGoLeft()) {
            this.pika.facing = this.LEFT;
            this.pika.x -= 1;
            break;
          }
        } else if (priority === this.UP) {
          if (this.pikachuCanGoUp()) {
            this.pika.facing = this.UP;
            this.pika.y -= 1;
            break;
          }
        }
      }

      if (this.pika.isTrapped()) {
        break;
      }
      this.rows[this.pika.y][this.pika.x] = (
        parseInt(this.rows[this.pika.y][this.pika.x]) + 1
      ).toString();
    }
  }

  public isPikachuBackAtTheStart(): boolean {
    return (
      this.pika.x === this.pika.start[0] &&
      this.pika.y === this.pika.start[1] &&
      this.rows[this.pika.y][this.pika.x] !== "0"
    );
  }

  public pikachuCanGoDown(): boolean {
    return (
      this.pika.y < this.height - 1 &&
      this.rows[this.pika.y + 1][this.pika.x] !== this.WALL
    );
  }

  public pikachuCanGoLeft(): boolean {
    return this.pika.x > 0 && this.rows[this.pika.y][this.pika.x - 1] !== this.WALL;
  }

  public pikachuCanGoRight(): boolean {
    return (
      this.pika.x < this.width - 1 &&
      this.rows[this.pika.y][this.pika.x + 1] !== this.WALL
    );
  }

  public pikachuCanGoUp(): boolean {
    return this.pika.y > 0 && this.rows[this.pika.y - 1][this.pika.x] !== this.WALL;
  }

  public printSolution(): void {
    for (const row of this.rows) {
      console.log(row.join(""));
    }
  }
}

const game = new Game();
game.readInput();
game.solve();
game.printSolution();
